Foxtrot/Driver/Rewriting/EmitAsyncClosure.cs

```diff
@@ -0,0 +1,969 @@
+                {
+                    this.forwarder.Visit(e);
+                }
+
+                ReplaceResult repResult = new ReplaceResult(
+                    this.checkPostMethod, this.originalResultLocal,
+                    this.rewriter.AssemblyBeingRewritten);
+
+                repResult.Visit(e);
+
+                if (repResult.ContractResultWasCapturedInStaticContext)
+                {
+                    this.contractResultCapturedInStaticContext.Add(e.Assertion.SourceContext);
+                }
+
+                // now need to initialize closure result fields
+                foreach (var target in repResult.NecessaryResultInitializationAsync(this.closureLocals))
+                {
+                    // note: target here 
+                    methodBody.Add(new AssignmentStatement(target, this.originalResultLocal));
+                }
+            }
+
+            // Emit normal postconditions
+
+            SourceContext? lastEnsuresSourceContext = null;
+
+            var ensuresChecks = new StatementList();
+
+            Method contractEnsuresMethod = this.rewriter.RuntimeContracts.EnsuresMethod;
+
+            foreach (Ensures e in GetTaskResultBasedEnsures(asyncPostconditions))
+            {
+                // TODO: Not sure that 'break' is enough! It seems that this is possible
+                // only when something is broken, because normal postconditions 
+                // are using Contract.Result<T>() and this is possible only for
+                // generic tasks.
+                if (IsVoidTask()) break; // something is wrong in the original contract
+
+                lastEnsuresSourceContext = e.SourceContext;
+
+                //
+                // Call Contract.RewriterEnsures
+                //
+
+                ExpressionList args = new ExpressionList();
+                args.Add(e.PostCondition);
+
+                args.Add(e.UserMessage ?? Literal.Null);
+
+                args.Add(e.SourceConditionText ?? Literal.Null);
+
+                ensuresChecks.Add(
+                    new ExpressionStatement(
+                        new MethodCall(
+                            new MemberBinding(null, contractEnsuresMethod),
+                            args, NodeType.Call, SystemTypes.Void),
+                        e.SourceContext));
+            }
+
+            this.rewriter.CleanUpCodeCoverage.VisitStatementList(ensuresChecks);
+
+            //
+            // Normal postconditions
+            //
+
+            // Wrapping normal ensures into following if statement
+            // if (task.Status == TaskStatus.RanToCompletion)
+            // { postcondition check }
+            //
+            // Implementation of this stuff is a bit tricky because if-statements
+            // are inverse in the IL.
+            // Basically, we need to generate following code:
+            // if (!(task.Status == Task.Status.RanToCompletion))
+            //   goto EndOfNormalPreconditions;
```

Comment:
:bulb: You use 'precondition' many times from here through where exceptional postconditions section starts. Do you mean to use 'postcondition' instead?